Data download script (for task purpose)

3Mydlo3 opened the pull request:
Based on:
https://github.com/Theseus-Aegis/Mods/pull/609

**When merged this pull request will:**
- Add data download script
  - Start the download via ACE action.
  - Display progress on laptop screen
  - Completes (fires `successful` event) when player unplugs the device (ACE action) after download reaches 100%
  - Public events for custom integrations:
    - `afmf_task_download_start` (server side only)
    - `afmf_task_download_started`
    - `afmf_task_download_successful`
    - `afmf_task_download_failed`

Requires ACE, skips loading if not loaded.

3Mydlo3 committed:
Initial version of task_download component

Some TODOs in the loop, but might not be needed at this moment

3Mydlo3 committed:
Add new HEMTT and .gitignore

3Mydlo3 committed:
Show 0% and 100% progress instead of skipping them

3Mydlo3 committed:
Add action to check download status

3Mydlo3 committed:
Complete download when player checks status once done

3Mydlo3 committed:
Add ace_interact_menu as dependency

3Mydlo3 committed:
Fix return type

3Mydlo3 committed:
Borrow display progress on screen from Theseus Mods

veteran29 commented on addons/task_download/XEH_postInit.sqf:
```diff
@@ -0,0 +1,37 @@
+    }] call CBA_fnc_addEventHandler;
+
+    [QGVAR(initCheckProgressAction), FUNC(initCheckProgressAction)] call CBA_fnc_addEventHandler;
+};
```
Event handlers should be added in PreInit if they're supposed to be working for JIP.

3Mydlo3 replied:
Removed that one.

veteran29 commented on addons/task_download/config.cpp:
```diff
@@ -0,0 +1,19 @@
+        requiredVersion = REQUIRED_VERSION;
+        requiredAddons[] = {
+            "afmf_main",
+            "ace_interact_menu"
```
We do not want hard dependency on ACE in this addon. Only CBA allowed, add `skipWhenMissingDependencies = 1;` or handle in script.

3Mydlo3 committed:
Add display of estimated time left and decrease interval to 1 s

veteran29 commented on addons/task_download/functions/fnc_setup.sqf:
```diff
@@ -0,0 +1,39 @@
+ * Example:
+ * [laptop] call afmf_task_download_fnc_setup
+ *
+ * Public: No
```
```suggestion
 * Public: Yes
```

veteran29 commented on addons/task_download/stringtable.xml:
```diff
@@ -0,0 +1,37 @@
+    <Package name="Task_Download">
+        <Key ID="STR_AFMF_Task_Download_DisplayName">
+            <English>Task Download</English>
+            <Polish>Zadane pobierania</Polish>
```
```suggestion
            <Polish>Zadanie pobierania</Polish>
```

veteran29 commented on addons/task_download/functions/fnc_started.sqf:
```diff
@@ -0,0 +1,28 @@
+    WARNING_1("Could not remove action from object %1",_object)
+};
+
+[_object, ACTION_TYPE, _actionPath] call ACEFUNC(interact_menu,removeActionFromObject);
```
I do not see much need for this as the condition should handle the action showing or not.

veteran29 commented on addons/task_download/functions/fnc_initCheckProgressAction.sqf:
```diff
@@ -0,0 +1,53 @@
+#include "script_component.hpp"
+/*
+ * Author: 3Mydlo3
+ * Initializes download action on given object.
```
```suggestion
 * Initializes check progress action on given object.
```

veteran29 reviewed (changes requested):
Feeling like this is quite more complicated than it needs to be.

ACE actions should basically be always added and show or not depending on their conditions, no much need to dynamically add/remove them, they do not cost much performance.

Also the download handling feels bit meh, I would just make it based on time instead of ticks as it basically allows every client to handle everything locally if they know start time (and that's all you need to broadcast then). I think the random time per tick does not brinch much value. If you want some randomness you can have some minor random coefficient for total download time that's not known.

3Mydlo3 committed:
Apply suggestions from code review

Co-authored-by: Filip Maciejewski <veteran29@users.noreply.github.com>

3Mydlo3 committed:
Merge remote-tracking branch 'origin/feature/task-download' into ddd

3Mydlo3 committed:
Simplify this

3Mydlo3 committed:
Last tweaks

3Mydlo3 committed:
Cleanup

3Mydlo3 committed:
Terminal lines fixes

3Mydlo3 committed:
Drop random date time

3Mydlo3 committed:
Last cleanup

3Mydlo3 commented:
> Feeling like this is quite more complicated than it needs to be.

I've simplified it heavily. It's now mostly clientside, except for download start & end. I've also cleaned up setup.

3Mydlo3 committed:
Delete unused stringtable entries

3Mydlo3 committed:
Update README

veteran29 commented on addons/task_download/functions/fnc_canContinue.sqf:
```diff
@@ -0,0 +1,24 @@
+if (isNull _object) exitWith { false };
+if (!alive _object) exitWith { false };
+
+true
```
Setting default value in params instead of doing `isNil` is IMO more idiomatic for SQF, null objects are not alive, so this can be greatly simplified:
```suggestion
params [["_object", objNull]];

alive _object // return
```

veteran29 committed:
Bump required arma version

veteran29 commented on addons/task_download/functions/fnc_formatEstimatedTimeRemaining.sqf:
```diff
@@ -0,0 +1,44 @@
+#include "script_component.hpp"
+/*
+ * Author: 3Mydlo3
+ * Function checks if download from given object can continue.
```
```suggestion
 * Formats remaining estimated time as a formatted string.
```

veteran29 commented on addons/task_download/functions/fnc_formatEstimatedTimeRemaining.sqf:
```diff
@@ -0,0 +1,44 @@
+ * 1: Current download progress <NUMBER> (Optional, default 0)
+ *
+ * Return Value:
+ * True if download can continue <STRING>
```
```suggestion
 * Formatted time <STRING>
```

veteran29 commented on addons/task_download/functions/fnc_formatEstimatedTimeRemaining.sqf:
```diff
@@ -0,0 +1,44 @@
+ * True if download can continue <STRING>
+ *
+ * Example:
+ * [laptop] call afmf_task_download_fnc_canContinue
```
```suggestion
 * [60, 50] call afmf_task_download_fnc_formatEstimatedTimeRemaining
```

veteran29 commented on addons/task_download/functions/fnc_setupSimple.sqf:
```diff
@@ -0,0 +1,53 @@
+ * Set up finished correctly <BOOL>
+ *
+ * Example:
+ * [laptop] call afmf_task_download_fnc_setupSimple
```
```suggestion
 * [laptop, 66, 2096] call afmf_task_download_fnc_setupSimple
```

veteran29 committed:
Fix missing public variable

veteran29 committed:
Rename _texureSource to _textureSelection

It's a hidden selection index or name.

veteran29 committed:
Better handling for object deletion

If we default the unplug value to true but set it on the server during init then `true` will be returned for objNull and this will prevent infinite loop.

veteran29 committed:
Add unplugged log

veteran29 committed:
Apply suggestions from code review

veteran29 committed:
Update example

veteran29 reviewed (approved):
LGTM

3Mydlo3 committed:
README example update
